### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by sloppyfocus**
This change fixes a number of bugs and adds tests to clustohttp:
- Handle add/set/delete attrs with Falsey values like 0
- Change urlencode to quote

### CODE REVIEW

File: test/test_clustohttp.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+        obj = obj.set_attr('foo', 'bar', 'garply')
+        new_attr = obj.attrs()[0]
+        self.assertEqual(len(obj.attrs()), 1)
+        self.assertEqual(new_attr['key'], 'foo')
+        self.assertEqual(new_attr['subkey'], 'bar')
+        self.assertEqual(new_attr['value'], 'garply')
+
+    def test_add_attr_zeroes(self):
+        c = get_mock_clusto()
```
Comment (shirchen) [inline test/test_clustohttp.py]: why didn't you make this an instance property in setUp?

File: test/test_clustohttp.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        c = get_mock_clusto(url=None)
+        self.assertEqual(c.url, 'http://testval')
+        del os.environ['CLUSTO_URL']
+
+    def test_no_environment_no_url(self):
+        with self.assertRaises(ValueError):
+            get_mock_clusto(url=None)
+
+    def test_request(self):
+        self.clusto.get_by_name('server01')
```
Comment (shirchen) [inline test/test_clustohttp.py]: any asserts on these?

File: test/test_clustohttp.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        new_attr = self._attr_from_params(query_params)
+
+        for attr in self.topology['server'][servername]['attrs']:
+            if attr['key'] == new_attr['key'] and attr['subkey'] == \
+                    new_attr['subkey']:
+                attr['value'] = new_attr['value']
+        return 200, {}, self.serialize_clusto_object(
+            servername, 'server', self.topology['server'][servername])
+
+    def __call__(self, method, path, body='', headers=None):
```
Comment (shirchen) [inline test/test_clustohttp.py]: setting mutable default args is frowned upon.

File: clustohttp.py [hunk 4]
```diff
@@ -282,9 +282,9 @@ def add_attr(self, key, subkey, value, datatype=None, number=None):
         Returns EntityProxy of the object the attribute added
         '''
         path = "addattr?" + "key=" + key + "&subkey=" + subkey + "&value=" + value
-        if datatype:
+        if datatype is not None:
```
Comment (shirchen) [inline clustohttp.py]: maybe comment on why we want empty data structs to slip though?

File: test/test_clustohttp.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
+import clustohttp
```
Comment (shirchen) [inline test/test_clustohttp.py]: :nit these should be organized according to PEP8: system/standard imports first, 3rd party, and then local

File: test/test_clustohttp.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        for clusto_type, clusto_objects in self.topology.items():
+            if name in clusto_objects:
+                response = self.serialize_clusto_object(name, clusto_type,
+                                                        clusto_objects[name])
+                response = "[" + response + "]"
+                return 200, {}, response
+        return 404, {}, '404 Not Found'
+
+    def get_entities(self, method, path, body, headers, query_params):
+        # TODO: Return a response based on self.topology
```
Comment (shirchen) [inline test/test_clustohttp.py]: TODO in this diff or later?

File: test/test_clustohttp.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        return 200, {}, '[]'
+
+    def _attr_from_params(self, params):
+        new_attr = {}
+        for k in ('key', 'subkey', 'value', 'datatype', 'number'):
+            if k in params:
+                new_attr[k] = params[k][0]
```
Comment (shirchen) [inline test/test_clustohttp.py]: list could be empty?

Comment (shirchen) [inline test/test_clustohttp.py]: also, is 87 an else of statement on 88? i would clarify.

File: test/test_clustohttp.py (hunk 5 above)
Comment (sloppyfocus) [inline test/test_clustohttp.py]: diddit

File: test/test_clustohttp.py (hunk 3 above)
Comment (sloppyfocus) [inline test/test_clustohttp.py]: strings are immutable

File: clustohttp.py (hunk 4 above)
Comment (sloppyfocus) [inline clustohttp.py]: If you pass an empty list, you'll get a 500 back from the clusto http api, which is I think what should happen. We can add lots of crazy validation here, but I think that it has a default arg of None, so we're just checking whether we should pass the arg based on if the api caller has explicitly called it or not

**Comment by shirchen** [comment]
:+1:

**Pull request merged**